pkg/controllers/sync/controller.go

```diff
@@ -1186,3 +1155,140 @@ func (s *SyncController) reconcileClusterForCascadingDeletion(
 
 	return worker.StatusAllOK
 }
+
+func (s *SyncController) cascadingDeletionForFTC(
+	ctx context.Context,
+	ftc *fedcorev1a1.FederatedTypeConfig,
+	syncMap *sync.Map, wg *sync.WaitGroup,
+	clusterName string,
+) {
+	defer wg.Done()
+
+	targetGVK := ftc.GetSourceTypeGVK().String()
+	ctx, logger := logging.InjectLoggerValues(ctx, "targetGVK", targetGVK)
+
+	// get all managed cluster objects according to ftc type
+	clusterObjs, err := s.listManagedClusterObjects(ctx, ftc, clusterName)
+	if err != nil {
+		logger.Error(err, "Failed to list clusterObjects")
+		syncMap.Store(targetGVK, fmt.Sprintf("failed to list cluster objects for %s", targetGVK))
+		return
+	}
+	if len(clusterObjs) == 0 {
+		return
+	}
+
+	// cascading-delete managed cluster objects
+	var deleteFailedObjs []string
+	for i := range clusterObjs {
+		clusterObj := clusterObjs[i].DeepCopy()
+		namespacedName := fmt.Sprintf("%s/%s", clusterObj.GetNamespace(), clusterObj.GetName())
+		_, keyedLogger := logging.InjectLoggerValues(ctx, "namespacedName", namespacedName)
+
+		// get the corresponding federated object
+		fedObjName := naming.GenerateFederatedObjectName(clusterObj.GetName(), ftc.Name)
+		fedObj, err := s.fedAccessor.FederatedResource(common.QualifiedName{
+			Namespace: clusterObj.GetNamespace(),
+			Name:      fedObjName,
+		})
+		if err != nil {
+			keyedLogger.Error(err, "Failed to get corresponding federated object")
+			deleteFailedObjs = append(deleteFailedObjs, namespacedName)
+			continue
+		}
+		if fedObj == nil {
+			err := fmt.Errorf("the federated object %s is not found", fedObj.FederatedName())
+			keyedLogger.Error(err, "Failed to get corresponding federated object")
+			deleteFailedObjs = append(deleteFailedObjs, namespacedName)
+			continue
+		}
+
+		// execute the cascading-delete operation
+		dispatcher := dispatch.NewUnmanagedDispatcher(s.getClusterClient, fedObj.TargetGVR(), fedObj.TargetName(), s.metrics)
+		s.removeFromCluster(ctx, dispatcher, clusterName, fedObj, clusterObj, true)
+
+		dispatchOk, timeoutErr := dispatcher.Wait()
+		if timeoutErr != nil {
+			keyedLogger.Error(timeoutErr, "Cascading-delete cluster object timeout")
```

Comment:
Do we need to append to `deleteFailedObjs`?

Reply:
It is not needed here. If `timeoutErr != nil`, `dispatchOk` must be false, so `deleteFailedObjs` is added directly under the judgment statement of `dispatchOk`. 
Additional comments are added here to explain the relationship between timeoutErr and dispatchOk.